File: 121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */

#include <vector>
using   namespace std;
// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int *dp = new int[prices.size()];
        dp[0] = prices[0];
        int score = 0;
        for (int i = 1; i < prices.size(); ++i) {
            dp[i] = min(dp[i - 1], prices[i]);
            score = max(score, prices[i] - dp[i - 1]);
        }
        return score;
    }
};
// @lc code=end

File: Swift/121.best-time-to-buy-and-sell-stock.swift
/*
 * @lc app=leetcode id=121 lang=swift
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start

// Solution 1
// T.C. = O(N^2)
// class Solution {
//     func maxProfit(_ prices: [Int]) -> Int {
//         var max = 0
//         for i in 0..<prices.count-1 {
//             for j in 1 ..< prices.count {
//                 max = max(max, prices[j]-prices[i])
//             }
//         }
//         return max
//     }
// }

// Solution 2
// T.C. = O(N)
// class Solution {
//     func maxProfit(_ prices: [Int]) -> Int {
//         var priceBuy = prices[0]
//         var maxProfit = 0

//         for idx in 1..<prices.count {
//             if prices[idx] < priceBuy {
//                 priceBuy = prices[idx]
//             }
//             else {
//                 maxProfit = max(maxProfit, prices[idx]-priceBuy)
//             }
//         }

//         return maxProfit
//     }
// }
// @lc code=end

